global/scripts/debug.shlib

```diff
@@ -0,0 +1,272 @@
+# Description:
+#
+#   Function for printing debugging information on caught errors.
+#   It makes use of error trapping. To reset the trap, add
+#   end of the script using "trap - ERR".
+#
+# Input:
+#
+#   — line number
+#   - command ran
+#   - exit status
+
+trap 'debug_on_error $LINENO "$BASH_COMMAND" $?' ERR
+
+if [[ "$(type -t log_Err)" != "function" ]]
+then
+    source "$HCPPIPEDIR/global/scripts/log.shlib" "$@"
+fi
+
+
+# Description:
+#
+#   Function to turn on error trappimg
+
+debug_enable_trap()
+{
+    HCP_DEBUG_USETRAP=ON
+}
+
+# Description:
+#
+#   Function to turn on off error trappimg
+
+debug_disable_trap()
+{
+    HCP_DEBUG_USETRAP=OFF
+}
+
+# Description:
+#
+#   Function to turn on exit upon error trap
+
+debug_enable_exit()
+{
+    HCP_DEBUG_EXIT_ON_ERROR=ON
+}
+
+# Description:
+#
+#   Function to turn off exit upon error trap
+
+debug_disable_exit()
+{
+    HCP_DEBUG_EXIT_ON_ERROR=OFF
+}
+
+# Description:
+#
+#   Function to turn on output coloring
+
+debug_enable_color()
+{
+    HCP_DEBUG_COLOR=TRUE
+}
+
+# Description:
+#
+#   Function to turn off output coloring
+
+debug_disable_color()
+{
+    HCP_DEBUG_COLOR=FALSE
+}
+
+
+# Description:
+#
+#   Function for printing red debug output.
+#
+# Input:
+#
+#   — text to be printed
+
+debug_red_echo()
+{   
+    if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+        echo -e "\033[31m $* \033[0m"
+    else
+        echo "$*"
+    fi
+}
+
+
+# Description:
+#
+#   Function for printing green debug output.
+#
+# Input:
+#
+#   — text to be printed
+
+debug_green_echo()
+{   
+    if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+        echo -e "\033[32m $* \033[0m"
+    else
+        echo "$*"
+    fi
+}
+
+
+# Description:
+#
+#   Function for printing blue debug output.
+#
+# Input:
+#
+#   — text to be printed
+
+debug_blue_echo()
+{   
+    if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+        echo -e "\033[34m $* \033[0m"
+    else
+        echo "$*"
+    fi
+}
+
+
+debug_on_error()
+{    
+    if [ ! $HCP_DEBUG_USETRAP == 'OFF' ]; then        
+
+        local me=`basename "$0"`
+        local line
+        local startreport = ""
+        local endreport   = ""
+
+        # --- try to identify command if $command is not set
+
+        if [ -z "$command" ]; then
```

Comment:
When is `command` already set before the trap, and why would we trust it before the `BASH_COMMAND` special variable that gets passed in?  `command` is a generic enough name that it might be used by scripts (and not unset after it is used), which seems like it could make the error reporting message wrong.

Reply:
My mistake. I incorrectly recalled that it was used. Will remove.